shared/modules/sizeof.js

```diff
@@ -0,0 +1,108 @@
+// Copyright 2014 Andrei Karpushonak
+
+'use strict';
+
+const ECMA_SIZES = {
+  STRING: 2,
+  BOOLEAN: 4,
+  BYTES: 4,
+  NUMBER: 8,
+  Int8Array: 1,
+  Uint8Array: 1,
+  Uint8ClampedArray: 1,
+  Int16Array: 2,
+  Uint16Array: 2,
+  Int32Array: 4,
+  Uint32Array: 4,
+  Float32Array: 4,
+  Float64Array: 8,
+};
+
+const isNodePlatform =
+  typeof process === 'object' && typeof require === 'function';
+
+function allProperties(obj) {
+  const stringProperties = [];
+  for (const prop in obj) {
+    if (Object.hasOwn(obj, prop)) {
+      stringProperties.push(prop);
+    }
+  }
+  if (Object.getOwnPropertySymbols) {
+    const symbolProperties = Object.getOwnPropertySymbols(obj);
+    Array.prototype.push.apply(stringProperties, symbolProperties);
+  }
+  return stringProperties;
+}
+
+function sizeOfObject(state, object) {
+  if (object === null) {
+    return;
+  }
+
+  const properties = allProperties(object);
+  const calc = getCalculator(state);
+  for (let i = 0; i < properties.length; i++) {
+    const key = properties[i];
+    calc(key);
+    calc(object[key]);
+  }
+}
+
+function getCalculator(state) {
+  return function calculator(object) {
+    if (state.size > state.maxSize) {
+      throw new Error('object exceeded max size');
+    }
+    if (Buffer.isBuffer(object)) {
+      state.size += object.length;
+    }
+
+    const objectType = typeof object;
+    switch (objectType) {
+      case 'string':
+        // https://stackoverflow.com/questions/68789144/how-much-memory-do-v8-take-to-store-a-string/68791382#68791382
+        state.size += isNodePlatform
```

Comment:
not understanding this? part are we creating a string here?

Reply:
no, calculating the byte size in memory of the string based on it's length and what execution environment the script is running in